Fixes #22385: Update user-management doc for custom-roles (#539)
Normation/rudder-plugins: user-management/README.adoc

[user-management-plugin]
= User management

This plugin allows to get new capabilities for user managements:
- you can assign fined-grained permissions to user through roles,

> - user get a self-service UI to manage an API token with the rights than what permission they have,

Reviewer: ```suggestion
- users get a self-service UI to manage an API token to perform API calls with their personal account,
```

- and finally, an UI is added in Rudder `Administration` menu to help with user managements.

== Unitary permissions

Rudder is using a system of fined-grained permissions to gain access to some part of the UI and to be able to
execute actions through it or APIs.

A permission is composed of a type and a level:

* Type:  Indicates what kind of data will be displayed and/or can be set/updated by the user
* Level: Access level to be granted on the related type

=== Permission type

> Permission types are linked to Rudder functional domain, like "things related to node management and inventories"

Reviewer: ```suggestion
Permission types are linked to Rudder functional domains, like "things related to node management and inventories"
```

(identified by the permission type `node`) or "configure and view rules" (identified by the permission id `rule`).
Rudder plugins can bring new permission types, documented in the plugin, like `cve` (for the CVE plugin).

> The list of core permission is:

Reviewer: ```suggestion
The list of core permissions is:
```

* *administration*: permissions related to Rudder settings, plugin management, user management, etc.
* *compliance*: permissions to access compliance information (especially with APIs).
* *configuration*: permission to manage configuration objects. It is a super-set of `rule`, `group`, `directive`, `technique`, `parameter`.
* *deployer* and *validator*: permission linked to change validation.
* *deployment*: permission linked to manual policy generation.
* *directive*: permission to access, create, delete and configure directives.
* *group*: permission to access, create, delete and configure groups.
* *node*: permission to accept or refuse nodes, access inventory, change properties or settings.
* *parameter*: permission to access, create, delete and configure parameters.
* *rule*: permission to access, create, delete and configure rules.
* *technique*: permission to access, create, delete and configure techniques.
* *userAccount*: permission related to user information access or personal API token management.

> In addition to these domain-bound permissions, Rudder manage internally two special permissions:

Reviewer: ```suggestion
In addition to these domain-bound permissions, Rudder internally manages two special permissions:
```

> * `no rights` is a special permission that forbid access to everything. It is given when something not expected happens (like permission computation errors, or `user-management` plugin disabled) so that users don't get more rights than they should be allowed.

Reviewer: ```suggestion
* `no rights` is a special permission that forbids access to everything. It is given when something not expected happens (like permission computation errors, or `user-management` plugin disabled) so that users don't get more rights than they should be allowed.
```

> * `any rights` is a special permission given to the `administrator` special role defined below that gives access to everything, whatever the rights to manage it is.

Reviewer: ```suggestion
* `any rights` is a special permission given to the `administrator` special role defined below that gives access to everything, whatever the rights to manage it are.
```

=== Permission levels

Rudder defines standard access level linked to permissions that reflects the action allowed for that permission kind:

* *read* to read given items
* *write* can create/delete items
* *edit* can modify existing items
* *all* which is a shortcut for "read, write, edit".

=== Permission syntax

> A permission is defined by the syntax `type_level`:

Reviewer: ```suggestion
A permission is defined by the syntax `${type}_${level}`:
```

* *configuration_read* will give read access to the configuration (rule management, directives, etc)
* *node_all* will give access to all action related to nodes.

=== Correspondence with API permissions

In Rudder, each permission grants access to some REST API endpoint mapping the corresponding action.

== Roles

Permission can be grouped by `roles` to ease the management of user permissions. Rudder comes with some predefined
roles and you can create your own custom role to fit your needs or take into account permissions provided by plugins.

=== Pre-defined roles

==== Administrator

Rudder has a special `administrator` role. That role, the only one enabled without the `user-management` plugin, grants
access and modification rights to everything, including Rudder settings, plugin management, etc. Use it with care.

==== Other pre-defined roles

Rudder comes with a set of pre-defined roles corresponding to the common use cases for interacting with Rudder.

> They are provided for convenience, but if they don't fit your need you can define your own role (see following paragraph).

Reviewer: ```suggestion
They are provided for convenience, but if they don't fit your need you can define your own roles (see following paragraph).
```

|validator | Can access and act on compliance and validator part
|====

The precise permission set for each role is presented below:

.Permission for pre-defined roles
image::docs/images/details_rights.png[]

=== Custom roles

You can define custom roles as union set of any permission or other roles.
Custom roles are defined in the files `/opt/rudder/etc/rudder-users.xml` with the following syntax:

```
<authentications>
  <custom-roles>
      <role name="read-only-restricted" roles="node_read,rule_read" />
      <role name="read-only-extended" roles="read-only-restricted,configuration_read" />
      <role name="cve-access" roles="cve_read" />
      <role name="auditor" roles="cve-access, compliance" />
  </custom-roles>

  <!-- same behavior than for roles -->
</authentications>
```

We can see that:

> * a custom role has two parameter:

Reviewer: ```suggestion
* a custom role has two parameters:
```

** a `name`, which must be unique, can't use unitary permission syntax, and can't reuse an existing

> name from a pre-defined role (but you can define alias)

Reviewer: ```suggestion
name from a pre-defined role (but you can define an alias)
```

> ** `roles`, which is a comma separated list of unitary permissions or roles. More precisely, the list can be

Reviewer: ```suggestion
** `roles`, which are a comma separated list of unitary permissions or roles. More precisely, the list can be
```

a list of unitary permissions (`read-only-restricted`,`cve-access`), or a mix of unitary permissions and roles
(`read-only-extended`), or even other pre-defined or custom roles (`auditor`).

The permissions granted by the role is the union of all unitary permissions granted by each unitary permissions
or roles.

== User permissions

User permissions are defined

> as a set of all fine-grained permissions assigned to him, either by role or

Reviewer: or?

== User management plugin interface

image::docs/images/usermanagement-display.png[]

image::docs/images/usermanagement-ui.png[]
image::docs/images/usermanagement-ui-api-token.png[]
image::docs/images/usermanagement-ui-user-details.png[]

WARNING: No Change Request will be generated when changes are made. The plugin is not subject to validation workflow when Change validation plugin is enabled.

> == Pre-defined roles

Reviewer: ```suggestion

== Pre-defined roles
```
